cdap-security/src/main/java/co/cask/cdap/security/authorization/AuthorizationUtil.java

```diff
@@ -160,7 +164,17 @@ public static String getAppAuthorizingUser(OwnerAdmin ownerAdmin, Authentication
     KerberosPrincipalId effectiveOwner =
       SecurityUtil.getEffectiveOwner(ownerAdmin, applicationId.getNamespaceId(),
                                      appOwner == null ? null : appOwner.getPrincipal());
-    return effectiveOwner != null ? effectiveOwner.getPrincipal() : authenticationContext.getPrincipal().getName();
+
+    // CDAP-13154 If impersonation is configured for either the application or namespace the effective owner will be
+    // a kerberos principal which can have different form
+    // (refer: https://docs.oracle.com/cd/E21455_01/common/tutorials/kerberos_principal.html). For example it can be
+    // a complete principal name (alice/somehost.net@someREALM). For authorization we need the enforcement to happen
+    // on the username and not the complete principal. The user name is the shortname of the principal so return the
+    // shortname as authorizing user.
+    String appAuthorizingUser = effectiveOwner != null ?
+      new KerberosName(effectiveOwner.getPrincipal()).getShortName() : authenticationContext.getPrincipal().getName();
```

Comment:
don't we need to use getShortName() also for the else case?

Reply:
the principal is always shortname. The AuthenticationContext is populated by the sign-in where we support just username and not principal to sign in.